doc/src/documentation_style_guide.rst

```diff
@@ -0,0 +1,1280 @@
+documentation, regardless if it is public or private.
+
+Docstrings should contain information aimed at users of the function. Comments
+specific to the code or other notes that would only distract users should go in
+comments in the code, not in docstrings.
+
+Every docstring should have examples that show how the function works. Examples
+are the most important part of a docstring. A single example showing input and
+output to a function can be more helpful than a paragraph of descriptive text.
+
+Remember that the primary consumers of docstrings are other human beings, not
+machines, so it is important to describe what the function does in plain
+English. Likewise, examples of how to use the function should be designed for
+human readers, not just for the doctest machinery.
+
+Keep in mind that while Sphinx is the primary way users consume docstrings, and
+therefore the first platform to keep in mind while writing docstrings
+(especially for public functions), it is not the only way users consume
+docstrings. You can also view docstrings using ``help()`` or ``?`` in IPython.
+When using ``help()``, for instance, it will show you all of the docstrings on
+private methods. Additionally, anyone reading the source code directly will see
+every docstring.
+
+All public functions, classes, and methods and their corresponding docstrings
+should be imported into the Sphinx docs, instructions on which can be found at
+the end of this guide.
+
+Docstring Formatting
+--------------------
+
+Docstrings are are written in `reStructuredText
+<http://docutils.sourceforge.net/rst.html>`_ format extended by `Sphinx
+<http://www.sphinx-doc.org/en/master/>`_. Here is a concise guide to `Quick
+reStructuredText <http://docutils.sourceforge.net/docs/user/rst/quickref.html>`_. More in-depth
+information about using reStructuredText can be found in the `Sphinx
+Documentation
+<http://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html>`_.
+
+In order for Sphinx to render docstrings nicely in the HTML documentation, some
+formatting guidelines should be followed when writing docstrings:
+
+* Always use """triple double quotes""" around docstrings. Use r"""raw triple
+  double quotes""" if you use any backslashes in your docstrings.
+* Include a blank line before the docstring’s closing quotes.
+* Lines should not be longer than 80 characters.
+* Always write class-level docstrings under the class definition line, as that
+  is better to read in the source code.
+* The various methods on the class can be mentioned in the docstring or
+  examples if they are important, but details on them should go in the
+  docstring for the method itself.
+* Be aware that :: creates code blocks, which should not be used in the
+  docstrings.
+* In order to make section underlining work nicely in docstrings, `numpydoc
+  Sphinx extension <https://pypi.org/project/numpydoc/>`_ is used. (WORK IN
+  PROGRESS: this may change via issues `#17599
```

Comment:
This should be updated. Those issues likely won't be addressed before this is merged.

Reply:
Ok I took out this WIP note.